File: frontend/src/services/society.ts
import api from "../lib/api";

export type SocietyRole = "leader" | "deputy_leader" | "secretary" | "treasurer" | "member";

// Matches db.ListSocietiesByYearRow
export interface Society {
  id: string;
  name: string;
  teacher_in_charge_id: string;
  academic_year_id: string;
  created_at: string;
  teacher_name: string;
  member_count: number;
}

// Matches db.Society — returned by GET /me/teacher/society, which has no
// join to teacher_profiles (the caller already knows their own name).
export interface SocietyBase {
  id: string;
  name: string;
  teacher_in_charge_id: string;
  academic_year_id: string;
  created_at: string;
}

// Matches db.ListSocietyYearsRow — academic years that have at least one
// society, for the archive view's year selector.
export interface SocietyYear {
  id: string;
  label: string;
  start_date: string;
}

// Matches db.ListSocietyMembersBySocietyRow
export interface SocietyMember {
  id: string;
  society_id: string;
  student_id: string;
  role: SocietyRole;
  academic_year_id: string;
  created_at: string;
  student_name: string;
  student_index: string;
  grade_name: string | null;
}

// Matches db.ListSocietyMembershipsByStudentRow — for the student portfolio.
export interface SocietyMembership {
  id: string;
  society_id: string;
  role: SocietyRole;
  academic_year_id: string;
  created_at: string;
  society_name: string;
  academic_year_label: string;
}

export interface CreateSocietyRequest {
  name: string;
  teacher_in_charge_id: string;
  academic_year_id: string;
}

export interface UpdateSocietyRequest {
  name: string;
  teacher_in_charge_id: string;
}

// The membership's academic year is always the society's own — the backend
// derives it server-side, so it isn't part of this request.
export interface AssignSocietyMemberRequest {
  student_id: string;
  role: SocietyRole;
}

export const societyApi = {
  listByYear: (academicYearId: string) =>
    api.get<Society[]>("/societies", { params: { academic_year_id: academicYearId } }).then((r) => r.data),

  listYears: () => api.get<SocietyYear[]>("/societies/years").then((r) => r.data),

  // The backend returns the raw db.Society row here (no teacher_name/
  // member_count join) — same shape as GET /me/teacher/society.
  create: (data: CreateSocietyRequest) => api.post<SocietyBase>("/societies", data).then((r) => r.data),

  update: (id: string, data: UpdateSocietyRequest) =>
    api.put<SocietyBase>(`/societies/${id}`, data).then((r) => r.data),

  remove: (id: string) => api.delete(`/societies/${id}`).then((r) => r.data),

  listMembers: (societyId: string) =>
    api.get<SocietyMember[]>(`/societies/${societyId}/members`).then((r) => r.data),

  assignMember: (societyId: string, data: AssignSocietyMemberRequest) =>
    api.put<SocietyMember>(`/societies/${societyId}/members`, data).then((r) => r.data),

  removeMember: (societyId: string, memberId: string) =>
    api.delete(`/societies/${societyId}/members/${memberId}`).then((r) => r.data),

  listByStudent: (studentId: string) =>
    api.get<SocietyMembership[]>(`/students/${studentId}/society-memberships`).then((r) => r.data),

  // the society the signed-in teacher is Teacher-in-Charge of, if any
  me: () => api.get<SocietyBase>("/me/teacher/society").then((r) => r.data),
};

File: backend/internal/models/society.go
package models

// CreateSocietyRequest creates a new society (club) for an academic year,
// admin-only. Rename or TIC reassignment happens via UpdateSocietyRequest.
type CreateSocietyRequest struct {
	Name              string `json:"name" binding:"required"`
	TeacherInChargeID string `json:"teacher_in_charge_id" binding:"required"`
	AcademicYearID    string `json:"academic_year_id" binding:"required"`
}

// UpdateSocietyRequest renames a society or reassigns its Teacher-in-Charge,
// admin-only. The academic year a society belongs to is immutable — create a
// new society for a new year instead, the same convention prefects/section
// heads use.
type UpdateSocietyRequest struct {
	Name              string `json:"name" binding:"required"`
	TeacherInChargeID string `json:"teacher_in_charge_id" binding:"required"`
}

// AssignSocietyMemberRequest appoints (or re-appoints, changing role) a
// student to a society's roster. The society is taken from the URL path,
// and the membership's academic_year_id is always the society's own year
// (SocietyService.AssignMember) — never client-supplied, so a caller can't
// create a membership row whose year doesn't match its society. Only the
// society's Teacher-in-Charge (or an admin) may call this — enforced in the
// service layer, not here. Role must be one of: leader, deputy_leader,
// secretary, treasurer, member.
type AssignSocietyMemberRequest struct {
	StudentID string `json:"student_id" binding:"required"`
	Role      string `json:"role" binding:"required,oneof=leader deputy_leader secretary treasurer member"`
}
